packages/sane/lib/src/impl/sane_sync.dart

```diff
@@ -0,0 +1,474 @@
+    }
+  }
+
+  @override
+  Future<void> dispose() {
+    if (_disposed) return Future.value();
+
+    final completer = Completer<void>();
+
+    Future(() {
+      _disposed = true;
+
+      dylib.sane_exit();
+      logger.finest('sane_exit()');
+
+      completer.complete();
+
+      _instance = null;
+    });
+
+    return completer.future;
+  }
+
+  @override
+  List<SyncSaneDevice> getDevices({required bool localOnly}) {
+    _checkIfDisposed();
+
+    final deviceListPointer =
+        ffi.calloc<ffi.Pointer<ffi.Pointer<SANE_Device>>>();
+
+    try {
+      final status = dylib.sane_get_devices(
+        deviceListPointer,
+        localOnly.asSaneBool,
+      );
+
+      logger.finest('sane_get_devices() -> ${status.name}');
+
+      status.check();
+
+      final devices = <SyncSaneDevice>[];
+
+      for (var i = 0; deviceListPointer.value[i] != ffi.nullptr; i++) {
+        final device = deviceListPointer.value[i].ref;
+        devices.add(SyncSaneDevice(device));
+      }
+
+      return List.unmodifiable(devices);
+    } finally {
+      ffi.calloc.free(deviceListPointer);
+    }
+  }
+
+  @pragma('vm:prefer-inline')
+  void _checkIfDisposed() {
+    if (_disposed) throw SaneDisposedError();
+  }
+}
+
+class SyncSaneDevice implements SaneDevice, ffi.Finalizable {
+  factory SyncSaneDevice(SANE_Device device) {
+    final vendor = device.vendor.toDartString();
+    return SyncSaneDevice._(
+      name: device.name.toDartString(),
+      vendor: vendor == 'Noname' ? null : vendor,
+      type: device.type.toDartString(),
+      model: device.model.toDartString(),
+    );
+  }
+
+  SyncSaneDevice._({
+    required this.name,
+    required this.vendor,
+    required this.model,
+    required this.type,
+  });
+
+  static final _finalizer = ffi.NativeFinalizer(dylib.addresses.sane_close);
+
+  SANE_Handle? _handle;
+
+  bool _closed = false;
+
+  @override
+  final String name;
+
+  @override
+  final String type;
+
+  @override
+  final String? vendor;
+
+  @override
+  final String model;
+
+  @override
+  void cancel() {
+    _checkIfDisposed();
+
+    final handle = _handle;
+
+    if (handle == null) return;
+
+    dylib.sane_cancel(handle);
+  }
+
+  SANE_Handle _open() {
+    final namePointer = name.toSaneString();
+    final handlePointer = ffi.calloc.allocate<SANE_Handle>(
+      ffi.sizeOf<SANE_Handle>(),
+    );
+
+    try {
+      dylib.sane_open(namePointer, handlePointer).check();
+      final handle = handlePointer.value;
+      _finalizer.attach(this, handle);
+      return handle;
+    } finally {
+      ffi.calloc.free(namePointer);
+      ffi.calloc.free(handlePointer);
+    }
+  }
+
+  @override
+  void close() {
+    if (_closed) return;
+
+    _closed = true;
+
+    if (_handle == null) return;
+
+    _finalizer.detach(this);
+    dylib.sane_close(_handle!);
+  }
+
+  @override
+  Uint8List read({required int bufferSize}) {
+    _checkIfDisposed();
+
+    final handle = _handle ??= _open();
```

Comment:
Sorry I totally forgotten your PR... too busy 😓

What I mean is to remove this line duplicated in almost each function:

```diff
- final handle = _handle ??= _open();
- dylib.sane_function(handle);
+ dylib.sane_function(_handleOrOpen);
```
The goal is just to make the code less verbose.

And yes, this could be merged with open, or inside a dedicated getter/method 🤔